main/src/ca/uwaterloo/flix/api/Bootstrap.scala

```diff
@@ -430,6 +439,102 @@ class Bootstrap(val projectPath: Path, apiKey: Option[String]) {
     }
   }
 
+  def lockEffects(flix: Flix): Result[Unit, BootstrapError] = {
+    Steps.updateStaleSources(flix)
+    for {
+      root <- Steps.check(flix)
+    } yield {
+      val json = EffectLocking.serialize(root)
+      val path = Bootstrap.getEffectLockFile(projectPath)
+      // N.B.: Do not use FileOps.writeJSON, since we uZse custom serialization formats.
+      FileOps.writeString(path, json)
+    }
+  }
+
+  /** Helper object for effect locking. */
+  private object EffectLocking {
+
+    /**
+      * Serializes the relevant functions for effect locking in `root` and returns a JSON string.
+      * The result may be written directly to a file.
```

Comment:
But it returns a string? I am confused.

Also, what if there is an IO error?

Reply:
There can be a serialization error but no IO error. We can wrap it in a `try-catch` and convert to ICE but there should not be an error if the project type checks.